File: src/contests/cg/puzzle/lua/mirrors.lua
-- 🎮 CodinGame Puzzle - mirrors
-- https://www.codingame.com/training/easy/mirrors

io.read()
local r = {}
for num in io.read():gmatch("%S+") do
  r[#r + 1] = tonumber(num)
end
local reflected = 0.0
for i = #r, 1, -1 do
  local ri = r[i]
  local denom = 1 - ri * reflected
  if denom == 0 then
    reflected = ri
  else
    reflected = ri + ((1 - ri) * (1 - ri) * reflected) / denom
  end
end
print(string.format("%.4f", reflected))

File: src/contests/cg/puzzle/haskell/mirrors.hs
-- 🎮 CodinGame Puzzle - mirrors
-- https://www.codingame.com/training/easy/mirrors

import Text.Printf (printf)

main :: IO ()
main = do
  _ <- getLine
  line <- getLine
  let r = map read (words line) :: [Double]
      step reflected ri =
        let denom = 1 - ri * reflected
         in ri + (if denom == 0 then 0 else ((1 - ri) * (1 - ri) * reflected) / denom)
      reflected = foldl step 0.0 (reverse r)
  printf "%.4f\n" reflected
